In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from IPython.display import display
pd.options.display.max_columns = None
pd.options.display.max_rows = None

In [8]:
df = pd.read_csv('/Users/sanjayroberts1/Desktop/data_science_tools1/sudeste.csv')

In [9]:
df = df.set_index('mdct')
df.index = pd.to_datetime(df.index)

In [8]:
d = {}
for col in df:
    d[col] = df[col].unique().tolist()

In [21]:
#print(d['temp'])

In [25]:
df.shape

(9779168, 30)

In [20]:
print(df.isna().sum())

wsid          0
wsnm          0
elvt          0
lat           0
lon           0
inme          0
city          0
prov          0
date          0
yr            0
mo            0
da            0
hr            0
prcp    8371184
stp           0
smax          0
smin          0
gbrd    4108820
temp         31
dewp        475
tmax         26
dmax        310
tmin         34
dmin        807
hmdy          0
hmax         12
hmin         44
wdsp     925561
wdct          0
gust     316474
dtype: int64


Some stations do not have wind or rain sensors, owing to why there are so many Null values. Will set them to zero to be able to do cleanup, then separate stations to do individual analysis

Gbrd has null values for when the sun is down (little less than half the time which makes sense). Make these zero values also to not affect analysis.

Other column null values we can set to zero and interpolate to get a value.

- Separate data for each station
- Each station may have it's own issues for data cleanup, so we separate and apply a cleanup function for each

In [22]:
df['prcp'].fillna(0,inplace=True)

In [31]:
df['gbrd'].fillna(0,inplace=True)

In [33]:
df.fillna(0,inplace=True)

In [34]:
df.isna().sum()

wsid    0
wsnm    0
elvt    0
lat     0
lon     0
inme    0
city    0
prov    0
date    0
yr      0
mo      0
da      0
hr      0
prcp    0
stp     0
smax    0
smin    0
gbrd    0
temp    0
dewp    0
tmax    0
dmax    0
tmin    0
dmin    0
hmdy    0
hmax    0
hmin    0
wdsp    0
wdct    0
gust    0
dtype: int64

Drop where all sensor columns are 0

In [36]:
col = ['prcp', 'stp', 'smax', 'smin', 'gbrd', 'temp',
       'dewp', 'tmax', 'dmax', 'tmin', 'dmin', 'hmdy', 'hmax', 'hmin', 'wdsp',
       'wdct', 'gust']
df = df[(df[col] != 0).any(axis=1)]


In [37]:
df.shape

(9159199, 30)

9779168 - 9159199 = 619969 columns dropped

take care of all but wind data. Look closer into wind data issues

In [39]:
df['temp'].replace(0, np.nan, inplace=True)
df['temp'].interpolate('time', inplace=True, limit_direction='both')
df['dewp'].replace(0, np.nan, inplace=True)
df['dewp'].interpolate('time', inplace=True, limit_direction='both')
df['tmax'].replace(0, np.nan, inplace=True)
df['tmax'].interpolate('time', inplace=True, limit_direction='both')
df['tmin'].replace(0, np.nan, inplace=True)
df['tmin'].interpolate('time', inplace=True, limit_direction='both')
df['dmax'].replace(0, np.nan, inplace=True)
df['dmax'].interpolate('time', inplace=True, limit_direction='both')
df['dmin'].replace(0, np.nan, inplace=True)
df['dmin'].interpolate('time', inplace=True, limit_direction='both')
df['hmax'].replace(0, np.nan, inplace=True)
df['hmax'].interpolate('time', inplace=True, limit_direction='both')
df['hmin'].replace(0, np.nan, inplace=True)
df['hmin'].interpolate('time', inplace=True, limit_direction='both')

Count number of nonzeros in each column

In [40]:
df.astype(bool).sum(axis=0)

wsid    9159199
wsnm    9159199
elvt    9119489
lat     9119489
lon     9119489
inme    9159199
city    9159199
prov    9159199
date    9159199
yr      9159199
mo      9159199
da      9159199
hr      8778203
prcp     661305
stp     9077445
smax    9073730
smin    9073810
gbrd    4790858
temp    9159199
dewp    9159198
tmax    9159199
dmax    9159199
tmin    9159199
dmin    9159199
hmdy    9099495
hmax    9159199
hmin    9159199
wdsp    8025423
wdct    8908786
gust    8621514
dtype: int64

Count number of zeros in each column

In [44]:
(df == 0).astype(int).sum(axis=0)

wsid          0
wsnm          0
elvt      39710
lat       39710
lon       39710
inme          0
city          0
prov          0
date          0
yr            0
mo            0
da            0
hr       380996
prcp    8497894
stp       81754
smax      85469
smin      85389
gbrd    4368341
temp          0
dewp          1
tmax          0
dmax          0
tmin          0
dmin          0
hmdy          0
hmax          0
hmin          0
wdsp    1133776
wdct     250413
gust     537685
dtype: int64

In [43]:
df['hmdy'].replace(0, np.nan, inplace=True)
df['hmdy'].interpolate('time', inplace=True, limit_direction='both')

look into why elevation, lat, lon is zero (all same weather station?), same with smin, smax etc, and look at surrornding values to see if it makes sense

hr, prcp, and gbrd are ok to be 0

https://stackoverflow.com/questions/40660088/get-first-row-of-dataframe-in-python-pandas-based-on-criteria

then work on wind stuff to clean that data up

In [22]:
df.to_hdf('cleaning_up_df.hdf','mydata',mode='w')

In [4]:
df = pd.read_hdf('cleaning_up_df.hdf','mydata')

In [48]:
df[df.elvt == 0].index[0]

Timestamp('2008-07-25 18:00:00')

In [50]:
df[df.elvt == 0].head()

wsid   wsnm  elvt  lat  lon  inme     city prov  \
mdct                                                                   
2008-07-25 18:00:00   418  MOELA   0.0  0.0  0.0  ZZZ3  Guarujá   SP   
2008-07-25 19:00:00   418  MOELA   0.0  0.0  0.0  ZZZ3  Guarujá   SP   
2008-07-25 20:00:00   418  MOELA   0.0  0.0  0.0  ZZZ3  Guarujá   SP   
2008-07-25 21:00:00   418  MOELA   0.0  0.0  0.0  ZZZ3  Guarujá   SP   
2008-07-25 22:00:00   418  MOELA   0.0  0.0  0.0  ZZZ3  Guarujá   SP   

                           date    yr  mo  da  hr  prcp     stp    smax  \
mdct                                                                      
2008-07-25 18:00:00  2008-07-25  2008   7  25  18   0.0  1014.3  1014.3   
2008-07-25 19:00:00  2008-07-25  2008   7  25  19   0.0  1014.8  1014.8   
2008-07-25 20:00:00  2008-07-25  2008   7  25  20   0.0  1015.2  1015.2   
2008-07-25 21:00:00  2008-07-25  2008   7  25  21   0.0  1015.8  1015.8   
2008-07-25 22:00:00  2008-07-25  2008   7  25  22   0.0  1016.7  1016.7   

                       smin     gbrd  temp  dewp  tmax  dmax  tmin  dmin  \
mdct                                                                       
2008-07-25 18:00:00  1014.0  426.700  20.0  16.0  20.2  16.9  19.8  15.8   
2008-07-25 19:00:00  1014.2  167.094  19.9  15.3  20.0  16.3  19.8  15.3   
2008-07-25 20:00:00  1014.8   95.589  19.6  15.4  19.9  15.7  19.5  15.2   
2008-07-25 21:00:00  1015.2   12.199  19.3  15.4  19.6  15.9  19.2  15.2   
2008-07-25 22:00:00  1015.8    0.000  19.2  14.0  19.3  15.5  19.2  13.7   

                     hmdy  hmax  hmin  wdsp   wdct  gust  
mdct                                                      
2008-07-25 18:00:00  78.0  83.0  76.0   3.3  242.0   7.2  
2008-07-25 19:00:00  75.0  80.0  75.0   2.5  239.0   7.4  
2008-07-25 20:00:00  77.0  78.0  74.0   2.8  222.0   6.8  
2008-07-25 21:00:00  78.0  80.0  77.0   1.9  276.0   8.0  
2008-07-25 22:00:00  72.0  79.0  70.0   4.0   94.0   8.4

In [52]:
df[df.elvt == 0].count()

wsid    39710
wsnm    39710
elvt    39710
lat     39710
lon     39710
inme    39710
city    39710
prov    39710
date    39710
yr      39710
mo      39710
da      39710
hr      39710
prcp    39710
stp     39710
smax    39710
smin    39710
gbrd    39710
temp    39710
dewp    39710
tmax    39710
dmax    39710
tmin    39710
dmin    39710
hmdy    39710
hmax    39710
hmin    39710
wdsp    39710
wdct    39710
gust    39710
dtype: int64

This explains the 0 values. Maybe the metadata for the station did not get input correctly? Let's go to the internet and see what the average lat/lon and elevation is for Guarujá

https://en.wikipedia.org/wiki/Guaruj%C3%A1

It is a very small area (55 sq miles)
Elevation is 4 meters
lat lon is 23°59′37″S 46°15′23″W --> lat -23.993611, lon -46.256389
We can change this in our dataset to get a more accurate description

In [5]:
df['elvt'].replace(0, 4, inplace=True)
df['lat'].replace(0, -23.993611, inplace=True)
df['lon'].replace(0, -46.256389, inplace=True)

Let's look at what needs cleaning next

In [6]:
(df == 0).astype(int).sum(axis=0)

wsid          0
wsnm          0
elvt          0
lat           0
lon           0
inme          0
city          0
prov          0
date          0
yr            0
mo            0
da            0
hr       380996
prcp    8497894
stp       81754
smax      85469
smin      85389
gbrd    4368341
temp          0
dewp          1
tmax          0
dmax          0
tmin          0
dmin          0
hmdy          0
hmax          0
hmin          0
wdsp    1133776
wdct     250413
gust     537685
dtype: int64

Let's look at stp. It looks like a bunch of the stations just use the same Smax value as they do for stp...

In [13]:
stp_smax_equal = df[df['stp'] == df['smax']].count().stp
stp_smax_equal

3652525

3652525/9159199 = 0.398, so about 40% of the data has this issue... Does this warrant dropping one of the columns? Or is smax a good enough 

Uhoh, then the other 40% has stp = smin

In [16]:
stp_smin_equal = df[df['stp'] == df['smin']].count().stp
stp_smin_equal

3458158

What about when smax = smin?

In [17]:
smax_smin_equal = df[df['smax'] == df['smin']].count().stp
smax_smin_equal

92853

9159199 - (3652525 + 3458158 + 92853) = 1955663 rows have independent values for all 3.

How much of our analysis do we want to use stp, smin, smax? Or drop? Or go thru for each station to see the issues and cleanup individually?

In [24]:
df[df.wdct == 0].head(300)

wsid            wsnm   elvt        lat        lon  inme  \
mdct                                                                           
2008-05-02 22:00:00   178     SÃO GONÇALO  237.0  -6.835777 -38.311583  A333   
2008-05-07 04:00:00   178     SÃO GONÇALO  237.0  -6.835777 -38.311583  A333   
2008-05-08 06:00:00   178     SÃO GONÇALO  237.0  -6.835777 -38.311583  A333   
2008-05-14 09:00:00   178     SÃO GONÇALO  237.0  -6.835777 -38.311583  A333   
2008-05-17 00:00:00   178     SÃO GONÇALO  237.0  -6.835777 -38.311583  A333   
2008-05-19 23:00:00   178     SÃO GONÇALO  237.0  -6.835777 -38.311583  A333   
2011-09-27 12:00:00   178     SÃO GONÇALO  237.0  -6.835777 -38.311583  A333   
2007-01-25 03:00:00   303         VITÓRIA    9.0 -20.271094 -40.306069  A612   
2008-06-08 01:00:00   303         VITÓRIA    9.0 -20.271094 -40.306069  A612   
2013-09-25 15:00:00   303         VITÓRIA    9.0 -20.271094 -40.306069  A612   
2013-09-25 16:00:00   303         VITÓRIA    9.0 -20.271094 -40.306069  A612   
2014-07-18 14:00:00   303         VITÓRIA    9.0 -20.271094 -40.306069  A612   
2014-07-18 15:00:00   303         VITÓRIA    9.0 -20.271094 -40.306069  A612   
2014-07-18 16:00:00   303         VITÓRIA    9.0 -20.271094 -40.306069  A612   
2015-06-18 14:00:00   303         VITÓRIA    9.0 -20.271094 -40.306069  A612   
2013-04-12 12:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2013-04-12 17:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2013-04-13 12:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2013-04-16 15:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2013-05-16 12:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2013-05-20 12:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2013-05-21 17:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2013-11-27 14:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2013-11-28 17:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2013-11-30 17:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2013-12-01 17:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2013-12-02 13:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2013-12-02 14:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2014-04-25 14:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2014-07-05 12:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2014-07-07 17:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2015-11-27 15:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2016-02-04 13:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2016-02-05 13:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2016-02-05 14:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2016-02-05 16:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2016-02-07 12:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2016-03-30 16:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2016-04-02 12:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2016-04-12 13:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2016-04-13 12:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2016-04-13 13:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2016-04-13 15:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2016-09-19 10:00:00   304    SANTA TERESA  976.0 -19.988388 -40.579572  A613   
2009-10-29 18:00:00   305        LINHARES   38.0 -19.356923 -40.068680  A614   
2010-11-26 12:00:00   305        LINHARES   38.0 -19.356923 -40.068680  A614   
2011-09-28 16:00:00   305        LINHARES   38.0 -19.356923 -40.068680  A614   
2013-06-20 19:00:00   305        LINHARES   38.0 -19.356923 -40.068680  A614   
2015-06-20 17:00:00  

Looks like some stations (178) did not capture wind gust data, though they have wind direction and speed. Then some stations do not capture all 3 measurements

In [25]:
df[df.gust == 0].head(300)

wsid         wsnm   elvt       lat        lon  inme  \
mdct                                                                       
2007-11-06 09:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-11-14 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-11-17 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-11-18 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-11-21 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-11-22 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-11-22 22:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-11-23 22:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-11-26 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-11-27 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-11-28 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-11-29 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-11-30 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-01 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-02 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-03 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-04 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-05 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-06 21:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-07 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-08 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-09 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-10 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-11 09:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-12 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-13 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-14 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-17 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-18 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-19 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-20 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-21 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-22 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-22 21:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-23 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-24 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-25 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-26 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-27 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-28 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-29 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-29 22:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-29 23:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2007-12-31 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2008-01-01 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2008-01-02 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2008-01-03 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2008-01-04 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2008-01-05 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2008-01-06 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A333   
2008-01-07 10:00:00   178  SÃO GONÇALO  237.0 -6.835777 -38.311583  A

In [31]:
len(df.city.unique())

117

Split station and data and store in a dict. Then we can individually clean each set?

In [26]:
individual_stations = {}
for each in df.wsid.unique():
    is_ = df['wsid'] == each
    k = str(each)
    individual_stations[k] = df[is_]

In [27]:
individual_stations.keys()

dict_keys(['178', '303', '304', '305', '306', '307', '308', '309', '310', '311', '312', '313', '314', '315', '316', '317', '318', '319', '320', '321', '322', '323', '324', '325', '326', '327', '328', '329', '330', '331', '332', '333', '334', '335', '336', '337', '338', '339', '340', '341', '342', '343', '344', '345', '346', '347', '348', '349', '350', '351', '352', '353', '354', '355', '356', '357', '358', '359', '360', '361', '362', '363', '364', '365', '366', '367', '368', '369', '370', '371', '372', '373', '374', '375', '376', '377', '378', '379', '380', '381', '382', '383', '384', '385', '386', '387', '388', '390', '391', '392', '393', '394', '395', '396', '397', '398', '399', '400', '401', '402', '403', '404', '405', '406', '407', '408', '409', '410', '411', '412', '413', '414', '415', '416', '417', '418', '419', '420', '421', '422', '423'])